### PR TITLE
Overhaul configuration file parsing

### PR DESCRIPTION
**Pull request by desilinguist**
- Use a new custom parser class `SKLLConfigParser` based on `ConfigParser`.
- Move all of the configuration parsing code to `config.py` and out of `experiments.py`.
- Add a `validate()` method to the config parser that raises a `KeyError` if:
  - unrecognized options are specified
  - options are specified in more than one section
  - options are specified in the wrong section (#223)
- Add unit tests for the above validation checks
- Update tests to deal with the above changes
- Update `requirements_rtd.txt` to install the latest version of `ConfigParser` (v3.5.0b2)

### CODE REVIEW

**Comment by desilinguist** [comment]
@dan-blanchard please do look at this one too since it's a somewhat major change.

File: tests/test_input.py [hunk 1]
```diff
@@ -671,16 +643,91 @@ def test_config_parsing_bad_cv_folds():
...
+
+    yield check_config_parsing_key_error, config_path
+
+
+def test_config_parsing_duplicate_option():
+
+    train_dir = join(_my_dir, 'train')
+    output_dir = join(_my_dir, 'output')
+
+    # make a simple config file that has an invalid option
```
Comment (aoifecahill) [inline tests/test_input.py]: minor: comment should be updated to reflect that the problem with this config is that there is a duplicate key (that comes from the fill_in_config_options method)

**Comment by aoifecahill** [comment]
This looks good to me, I didn't spot anything obvious that might cause problems.

**Comment by dan-blanchard** [comment]
I did a little PEP8 clean up here and there, but otherwise this looks great! :+1:

**Comment by desilinguist** [comment]
Thanks, Dan! I will merge as soon as the build passes.

**Comment by dan-blanchard** [comment]
I actually already fixed that comment too. :)

**Comment by desilinguist** [comment]
Sweet! Thanks for that!

Actually, I just realized I didn't address #213 in this PR. Do you think that's important for this new release?

**Comment by dan-blanchard** [comment]
It would be nice to have fixed for this release, since we know it's bitten some new users when trying to follow the tutorial.

**Comment by desilinguist** [comment]
Okay, I will add that to this PR then and try to add some tests for it too.

**Comment by desilinguist** [comment]
@dan-blanchard and @aoifecahill, can you please take a look again? I just committed some changes to address #213.

File: skll/config.py [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+    if random_folds:
+        if cv_folds_file:
+            logger.warning('Specifying cv_folds_file overrides random_folds')
+        do_stratified_folds = False
+    else:
+        do_stratified_folds = True
+
+    # get all the input paths and directories (without trailing slashes)
+    train_path = config.get("Input", "train_directory").rstrip('/')
+    test_path = config.get("Input", "test_directory").rstrip('/')
```
Comment (dan-blanchard) [inline skll/config.py]: This won't work on Windows.

Comment (desilinguist) [inline skll/config.py]: This has been in there since the beginning so I guess it has never worked?

**Comment by desilinguist** [comment]
I replaced the `/` with `os.sep` which should work on Windows too.

File: tests/test_input.py [hunk 3]
```diff
@@ -671,16 +645,151 @@ def test_config_parsing_bad_cv_folds():
...
-                                                   'bad_cv_folds')
+                                                   'mislocated_input_file')
+
+    yield check_config_parsing_file_not_found_error, config_path
 
-    _parse_config_file(config_path)
+
+def test_config_parsing_relative_input_path():
+
+    train_dir = '../train'
```
Comment (dan-blanchard) [inline tests/test_input.py]: This should also be:

```python
train_dir = join('..', 'train')
```

Comment (desilinguist) [inline tests/test_input.py]: Ah, good call.

**Comment by desilinguist** [comment]
So, is this okay to be merged once I address your last comment?

**Comment by dan-blanchard** [comment]
Definitely.

**Pull request merged**